metaflow/plugins/airflow/airflow_utils.py

```diff
@@ -0,0 +1,410 @@
+import hashlib
+import json
+import os
+import random
+import re
+import time
+import typing
+from collections import defaultdict
+from datetime import datetime, timedelta
+
+
+class KubernetesProviderNotFound(Exception):
+    headline = "Kubernetes provider not found"
+
+
+class AirflowSensorNotFound(Exception):
+    headline = "Sensor package not found"
+
+
+LABEL_VALUE_REGEX = re.compile(r"^[a-zA-Z0-9]([a-zA-Z0-9\-\_\.]{0,61}[a-zA-Z0-9])?$")
+
+TASK_ID_XCOM_KEY = "metaflow_task_id"
+RUN_ID_LEN = 12
+TASK_ID_LEN = 8
+RUN_ID_PREFIX = "airflow"
+
+# AIRFLOW_TASK_ID_TEMPLATE_VALUE will work for linear/branched workflows.
+# ti.task_id is the stepname in metaflow code.
+# AIRFLOW_TASK_ID_TEMPLATE_VALUE uses a jinja filter called `task_id_creator` which helps
+# concatenate the string using a `/`. Since run-id will keep changing and stepname will be
+# the same task id will change. Since airflow doesn't encourage dynamic rewriting of dags
+# we can rename steps in a foreach with indexes (eg. `stepname-$index`) to create those steps.
+# Hence : Foreachs will require some special form of plumbing.
+# https://stackoverflow.com/questions/62962386/can-an-airflow-task-dynamically-generate-a-dag-at-runtime
+AIRFLOW_TASK_ID_TEMPLATE_VALUE = (
+    "%s-{{ [run_id, ti.task_id, dag_run.dag_id ] | task_id_creator  }}" % RUN_ID_PREFIX
+)
+
+
+class SensorNames:
+    EXTERNAL_TASK_SENSOR = "ExternalTaskSensor"
+    SQL_SENSOR = "SQLSensor"
+    S3_SENSOR = "S3KeySensor"
+
+    @classmethod
+    def get_supported_sensors(cls):
+        return list(cls.__dict__.values())
+
+
+def run_id_creator(val):
```

Comment:
can we combine `task_id_creator` and `run_id_creator`? They seem to be practically the same modulo length.

Reply:
I can remove the functions and make it one function but the airflow macro should ideally be different. `run_id_creator` is used like [this](https://github.com/valayDave/metaflow/blob/148549cc6adf18a1b7adeb0821f2002fdf7aeb39/metaflow/plugins/airflow/airflow.py#L341)